### PR TITLE
findprev: use zero(eltype(A)) instead of 0

### PR DESCRIPTION
**Pull request by Keno**

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
What was the issue here?

**Comment by Keno** [comment]
Some types don't implement comparison/promotion with Int64, or if they do, it's not what you want, e.g.

```
using SIUnits
julia> findlast([0m, 1m, 0m])
3
```
because
```
julia> 0m == 0
**false**
```
(arguable that should have been an error instead)

File: base/array.jl [hunk 1]
```diff
@@ -1111,7 +1111,7 @@ julia> findprev(A,1)
 """
 function findprev(A, start::Integer)
     for i = start:-1:1
-        A[i] != 0 && return i
+        A[i] != zero(eltype(A)) && return i
```
Comment (andreasnoack) [inline base/array.jl]: Might be worth making this `zero(A[i])` in case `eltype(A)=Any`

Comment (Keno) [inline base/array.jl]: Fair point.

**Comment by Keno** [comment]
Not sure what happened here. I neither merged nor closed this.

File: base/array.jl [hunk 2]
```diff
@@ -1111,15 +1111,15 @@ julia> findprev(A,1)
...
+        A[i] != zero(A[i]) && return i
     end
     return 0
 end
 
 """
     findlast(A)
 
-Return the linear index of the last non-zero value in `A` (determined by `A[i]!=0`).
+Return the linear index of the last non-zero value in `A` (determined by `A[i]!=zero(A[i])`).
```
Comment (tkelman) [inline base/array.jl]: needs a genstdlib run

**Comment by simonbyrne** [comment]
If we're going to try to support these sort of things it would be good to make a formal specification of the interfaces that need to be supported. As part of my current work I've been creating yet another unitful computation package, and so I've hit a few of these issues as well.

The current pattern is used in quite a few places throughout Base, as there are some cases where it is more efficient (e.g. `BigFloat` vs `Int` has an inequality comparison method which doesn't require allocating a new `BigFloat`)

**Comment by tkelman** [comment]
x-ref https://github.com/JuliaLang/julia/pull/15755

**Comment by stevengj** [comment]
Seems like we should define an `iszero(x) = x == zero(x)` function, with specialized versions for bignums and arrays that avoids allocations.

**Comment by nalimilan** [comment]
Are we sure we want `0m != 0` and yet `iszero(0m) == true`? That sounds potentially confusing to me. Types should either compare with other numbers the standard way, or not compare with them at all. Maybe only the definition of `find` should be changed?

**Comment by stevengj** [comment]
@nalimilan.  Yes.   `zero(x)` is the additive identity for `x`.  It is not (necessarily) `0`.   `iszero` should check whether something is the additive identity.  `zeros(5,5) == 0` is `false`, too.

**Comment by ararslan** [comment]
Out of curiosity, why change `findprev` and not the other `find*` functions?

**Comment by Keno** [comment]
Superseded by the find rework stuff.

**Pull request closed without merging**